### PR TITLE
[canvaskit] Optimize CanvasKit platform views in special cases

### PR DESCRIPTION
**Pull request by hterkelsen**
Diff the old and new platform view list and attempt to elminate most DOM manipulation in the common case.

Addresses https://github.com/flutter/flutter/issues/86868 for CanvasKit

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/surface_factory.dart [hunk 1]
```diff
@@ -29,6 +29,12 @@ class SurfaceFactory {
   /// The maximum number of surfaces which can be live at once.
   final int maximumSurfaces;
 
+  /// The maximum number of assignable overlays.
+  ///
+  /// This is just `maximumSurfaces - 2` (the maximum number of surfaces minus
+  /// the base surface and overlay surface).
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/surface_factory.dart:35]: nit: did you mean base surface and _backup_ surface?

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart [hunk 2]
```diff
@@ -718,3 +864,85 @@ class MutatorsStack extends Iterable<Mutator> {
...
+/// Diff the composition order with the active composition order. It is
+/// common for the composition order and active composition order to differ
+/// only slightly.
+///
+/// Consider a scrolling list of platform views; from frame
+/// to frame the composition order will change in one of two ways, depending
+/// on which direction the list is scrolling. One or more views will be added
+/// to the beginning of the list, and one or more views will be removed from
+/// the end of the list, with the order of the unchanged middle views
+/// remaining the same.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:901]: Checkout [_insertChildDomNodes](https://github.com/flutter/engine/blob/d877a4bbcd0e5b48d2ff18fcba7e010dc52276fe/lib/web_ui/lib/src/engine/html/surface.dart#L1035). With the same amount of code (reusing `longestIncreasingSubsequence`) we should be able to find the optimal solution for all situations, including the scroll case.

File: lib/web_ui/lib/src/engine/canvaskit/surface_factory.dart (hunk 1 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/surface_factory.dart:35]: Yes, done

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:901]: Added a TODO. That refactor would also necessitate copying more code, for instance to generate the index maps.

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**